### PR TITLE
Centralized web manager

### PR DESCRIPTION
**Pull request by tonyo**
This large PR adds some functionality for AD management. This is a follow-up to my presentation last Wednesday (see the [presentation](https://www.dropbox.com/s/992mn9653w5iwbr/ScionTestbed.pdf?dl=0) to get a high-level overview of the proposed changes).

Two directories were added to the package root:
1. `ad_management`
   
   This directory contains the code of the two important modules: the management daemon (`management_daemon.py`) and the updater (`updater.py`). The management daemon is managed by Supervisor, so make sure you are using the updated version of `supervisord.conf`.
2. `web_scion`
   
   Django web application is located here. See the README.md for setup and running instructions. There's a single app called `ad_manager` which does almost all the work.
   Not all the features are fully implemented: for example, there's still work to do in terms of update management, AD control, and security, among others.

### CODE REVIEW

File: ad_management/secure_rpc_server.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            lambda _, v, w: w("<value><i8>%d</i8></value>" % v)
+
+        socketserver.BaseServer.__init__(self, addr, requestHandler)
+        self.socket = ssl.wrap_socket(
+            socket.socket(self.address_family, self.socket_type),
+            server_side=True,
+            cert_reqs=ssl.CERT_NONE,  # TODO
+            certfile=os.path.join(CERT_DIR_PATH, 'cert.pem'),
+            keyfile=os.path.join(CERT_DIR_PATH, 'key.pem'),
+            ssl_version=ssl.PROTOCOL_SSLv23,
```
Comment (pszal) [inline ad_management/secure_rpc_server.py]: that sounds insecure

File: web_scion/ad_manager/migrations/0001_initial.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                ('id', models.CharField(max_length=50, serialize=False, primary_key=True)),
+            ],
+            options={
+            },
+            bases=(models.Model,),
+        ),
+        migrations.CreateModel(
+            name='ISD',
+            fields=[
+                ('id', models.CharField(max_length=50, serialize=False, primary_key=True)),
```
Comment (pszal) [inline web_scion/ad_manager/migrations/0001_initial.py:25]: unique ?

File: web_scion/ad_manager/models.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        super(SelectRelatedModelManager, self).__init__()
+        self.related_fields = args
+
+    def get_queryset(self):
+        queryset = super(SelectRelatedModelManager, self).get_queryset()
+        return queryset.select_related(*self.related_fields)
+
+
+class ISD(models.Model):
+    id = models.CharField(max_length=50, primary_key=True)
```
Comment (pszal) [inline web_scion/ad_manager/models.py:29]: unique

File: web_scion/ad_manager/models.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+    def __str__(self):
+        return str(self.id)
+
+    class Meta:
+        verbose_name = 'ISD'
+
+
+class AD(models.Model):
+    id = models.CharField(max_length=50, primary_key=True)
```
Comment (pszal) [inline web_scion/ad_manager/models.py:39]: maybe check uniqueness within ISD before save()

File: web_scion/ad_manager/views.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+from django.http import JsonResponse, HttpResponse, HttpResponseNotFound
+from django.shortcuts import redirect
+from django.views.generic import ListView, DetailView
+from ad_manager.models import AD, ISD
+from ad_manager.util import monitoring_client
+from ad_management.common import (is_success, get_success_data,
+    ARCHIVE_DIST_PATH, get_data, response_failure)
+from lib.topology import Topology
+
+ARCH_NAME = 'scion-0.1.0.tar.gz'
```
Comment (pszal) [inline web_scion/ad_manager/views.py:26]: hardcoded for testing?

File: ad_management/secure_rpc_server.py (hunk 1 above)
Comment (tonyo) [inline ad_management/secure_rpc_server.py]: I agree, will change it to `ssl.PROTOCOL_TLSv1_2`

File: web_scion/ad_manager/views.py (hunk 5 above)
Comment (tonyo) [inline web_scion/ad_manager/views.py:26]: Yes, it's not here anymore in later commits.

File: web_scion/ad_manager/models.py (hunk 3 above)
Comment (tonyo) [inline web_scion/ad_manager/models.py:29]: `primary_key=True` implies global uniqueness, if I understood you correctly.
By the way, is it allowed to have two ADs with the same id in two different ISDs?

Comment (tonyo) [inline web_scion/ad_manager/models.py:29]: As a sidenote, I changed the types of `ISD.id` and `AD.id` to integer some commits after this PR.

**Pull request merged**